### ISSUE TITLE
Fix bash installation on AIX

### ISSUE THREAD

**Issue by sxa** [adoptium/infrastructure#1320]
At the moment the version of bash on new AIX machines (the `test-ibm-` ones) is [causing problems](https://github.com/eclipse/openj9/issues/9369#issuecomment-624634760) when running some of the builds. It's hitting pipe failures. It's currently unclear whether this is `ulimit`-related (I did some quick tests but they were not conclusive) therefore I have backed out the version of bash on the AIX 7.1SP5 machine used for building OpenJ9 to the version that's on the old build machines, as per the final bullet point of https://github.com/AdoptOpenJDK/openjdk-build/issues/1733#issue-614108892. This is not repeatable in the playbooks so an alternate solution needs to be found, and the machine updated to reflect the solution we put in place if required.

**Comment by aixtools** [comment]
I have not looked at bash for quite awhile - so I will not say I have the latest, but I do offer an alturnative via: http://www.aixtools.net/index.php/bash

There you will find both bash4 and bash5 - and one of the two can be installed - in parallel - with the AIX Toolbox, Michael Perzl, or Freeware RPM packaged version.

Just need to call /opt/bin/bash rather than their bash.

Iirc, there are no other requirements - if there are - I'll repackage to remove them.

**Comment by aixtools** [comment]
p.s. I just built bash-5.0 patchlevel 18, and checked bash-4.4 patchlevel 23 - both have an error in (sub)test vredir6.sub.

Using an RPM based package - and I expect pre-patchlevel 23 - the error is still there:

BASH_VERSINFO=([0]="4" [1]="4" [2]="0" [3]="1" [4]="release"
[5]="powerpc-ibm-aix6.1.0.0")
BASH_VERSION='4.4.0(1)-release'
...

```

_=/usr/bin/bash
aixtools@gcc119:[/home/aixtools]PS1='$ '
$ unset v
$ exec {v}</dev/null
$ echo $v
10
$ exec {v}<&-
$ ulimit -n
2000
$ ulimit -n 6
$ ulimit -n
6
$ exec </dev/null
$ exit
aixtools@gcc119:[/home/aixtools]
```
The shell should not be responding with an `exit` after `exec </dev/null`

Researching...

**Comment by aixtools** [comment]
On 15/10/2020 09:39, Michael Felt wrote:
>
> p.s. I just built bash-5.0 patchlevel 18, and checked bash-4.4
> patchlevel 23 - both have an error in (sub)test vredir6.sub.
>
> Using an RPM based package - and I expect pre-patchlevel 23 - the
> error is still there:
>
> BASH_VERSINFO=([0]="4" [1]="4" [2]="0" [3]="1" [4]="release"
> [5]="powerpc-ibm-aix6.1.0.0")
> BASH_VERSION='4.4.0(1)-release'
> ...
>
> |_=/usr/bin/bash aixtools@gcc119:[/home/aixtools]PS1='$ ' $ unset v $
> exec {v}</dev/null $ echo $v 10 $ exec {v}<&- $ ulimit -n 2000 $
> ulimit -n 6 $ ulimit -n 6 $ exec </dev/null $ exit
> aixtools@gcc119:[/home/aixtools] |
>
> The shell should not be responding with an |exit| after |exec </dev/null|
>
|The reply from the bash group:|

|a) from the command-line `exec </dev/null` closes the current shell -
as stdin - was cli, and is unavailable.
When from from a script this is not the case, and the script continues.

The error message is yet another example of AIX using different
constants for error messages. It appears this test failure may be ignored.|

|b) I have the new packaging for bash - with no added dependencies -
available via: wiki: http://www.aixtools.net/index.php/bash, bff:
http://download.aixtools.net/tools/gnu/aixtools.gnu.bash.5.0.18.0.I

c) Soon - I hope after PR #1633 merges - I can add a role that can
manage to download and install via ansible.
|

**Comment by aixtools** [comment]
I cannot comment on whether my packaging of bash resolves the issue or not. Can someone check?

Here is a playbook that can be used to install 'aixtools' bash - to assist with checking if this bash resolves the issue - rather than re-installing a back-version of bash.

```
# Install bash when /opt/bin/bash does not exist
# expected to provide definition of target host using --extra-vars
- hosts: "{{ tgt_host | default('nohostname') }}"
  gather_facts: no

  vars:
    tools:  "http://download.aixtools.net/tools/gnu/"
    lpp:    "aixtools.gnu.bash"
    vrmf:   "5.0.18.0"
    chksum: "md5:E6106A3CF9A82B4E8785003F587ECA13"
    idem:   "/opt/bin/bash"

  tasks:
    - name: "Check for existing {{ lpp }}"
      stat:
        path: "{{ idem }}"
      register: tgt

    - name: "Download and Install {{ lpp }}"
      block:
        - get_url:
            url: "{{ tools }}{{ lpp }}.{{ vrmf }}.I"
            dest: "/tmp"
            checksum: "{{ chksum }}"
        - command: "/usr/sbin/installp -d /tmp/{{ lpp }}.{{ vrmf }}.I -aYX all"
          register: installp
        - file: "path=/tmp/{{ lpp }}.{{ vrmf }}.I state=absent"
        - file: "path=/tmp/.toc state=absent"
        - file:
            src: /opt/bin/bash
            dest: /usr/bin/bash
            state: link
        - debug:
            var: "{{ item }}.stdout_lines"
          with_items:
            - installp
      when: tgt.stat.exists == false
```

**Comment by aixtools** [comment]
iirc - the IBM hosted systems are running AIX 7.1 TL5 (SP1).

Question: is this still an issue on those systems - if yes, can it be replicated on test-osuosl-aix71-ppc64-1 - which is also AIX 7.1 TL5 (but SP2). If it can be replicated - then I can research further on the OSUOSL node.

**Comment by aixtools** [comment]
Is this still an issue, or can it be closed?

**Comment by aixtools** [comment]
The issue is a special function of bash being used, that opens pipes aka fifo files - that bash does not remove when it finishes.

Now that I am beginning to understand where fixes can be put - I'll try and come up with a fix that does not use this feature - in the way it is currently used - at least.

**Comment by aixtools** [comment]
I went looking for issues of bash using unnamed fifo redirects. That is what I found to be the `root cause` during earlier research.

In the make files there are re-directs using the bash `feature` >( ... ) and <( ... ).

During the testing phase I saw many examples of `>( tee ... )` - that are not really needed  `| tee ...` would work just as well.

However, in checking the build scripts I found one example in a `.gmk` source that might be difficult to re-write.

```
define generate-preproc-src
        $(call MakeDir, $(@D))
        $(call ExecuteWithLog, $(SUPPORT_OUTPUTDIR)/gensrc/java.base/_$(@F), \
            ( $(AWK) '/@@END_COPYRIGHT@@/{exit}1' $< && \
              $(CPP) $(CPP_FLAGS) $(SYSROOT_CFLAGS) $(CFLAGS_JDKLIB) $(CPP_FILEPREFIX) $< \
                  2> >($(GREP) -v '^$(<F)$$' >&2) \
                  | $(AWK) '/@@START_HERE@@/,0' \
                  |  $(SED) -e 's/@@START_HERE@@/\/\/ AUTOMATICALLY GENERATED FILE - DO NOT EDIT/' \
                  -e 's/PREFIX_//' -e 's/^#.*//' \
            ) > $@ \
        )
endef
```
The line `2> >($(GREP) -v '^$(<F)$$' >&2) \` is actually doing something other than just copying the output to a file. My gut feeling is that it can be rewritten to not need the unnamed FIFO as a replacement for unnamed pipe (`|`), but I do not recognize, directly what `$(<F)$$` actually is. A little too terse for me, sadly.

Anyone who can help me understand what is happening here (and especially, how).

**Comment by aixtools** [comment]
@sxa - as this started on the _ibm-test_ machines, and they no longer exist - is this still an issue?

Summary of previous findings:
bash has a special syntax to redirect input/output to sub-shells (using one, or both of `>( ... ) and <( ... )`.

The implementation creates so-called `fifo` (named) files in /tmp. AIX needs them to be specifically `unlink()`ed, unlike a pipe() which is not named.

If a feature change in `bash` attempted to do (more of) this in memory (with `fork()` used to copy the redirection - that may have been a concern.

- My suggestion is to close this ticket as no longer relevant (the systems no longer exist).

**Comment by sxa** [comment]
Good point - I skipped over this during my last sweep as I hadn't spotted it was specific to those, so yes I'm happy to close it.
